internal/stationxml/response.go

```diff
@@ -0,0 +1,468 @@
+// SetFrequency is used to set the overall reference frequency for the Response.
+func (r *Response) SetFrequency(frequency float64) {
+	Frequency(frequency)(r)
+}
+
+// SetCalibration is used to set a initial sensor reference gain, this overrides the default values.
+func (r *Response) SetCalibration(scale, bias float64) {
+	Calibration(scale, bias)(r)
+}
+
+// SetGain is used to adjusts the installed sensor gains, this is in addition to the default values.
+func (r *Response) SetGain(scale, bias, absolute float64) {
+	Gain(scale, bias, absolute)(r)
+}
+
+// SetPreamp is used to adjusts the datalogger gains, this is in addition to the default values.
+func (r *Response) SetPreamp(preamp float64) {
+	Preamp(preamp)(r)
+}
+
+// Polynomial finds the PolynomialType in the Response if one is present.
+func (r *Response) Polynomial() *PolynomialType {
+	for _, s := range r.stages {
+		if s.Polynomial == nil {
+			continue
+		}
+		return s.Polynomial
+	}
+	return nil
+}
+
+// SetSensor takes an XML encoded ResponseType that represents a Sensor and adds it to the Response.
+func (r *Response) SetSensor(data []byte) error {
+
+	var sensor ResponseType
+	if err := xml.Unmarshal(data, &sensor); err != nil {
+		return err
+	}
+
+	switch {
+	case sensor.InstrumentSensitivity != nil:
+		// check biases, as these imply a polynomial response
+		if r.ScaleBias != 0.0 || r.GainBias != 0.0 || r.GainAbsolute != 0.0 {
+			return ErrInvalidResponse
+		}
+		// a simple scaling of the overall sensitivity
+		sensor.InstrumentSensitivity.Value *= (r.ScaleFactor * r.GainFactor)
+		// look for the first stage with a gain
+		for i := range sensor.Stages {
+			stage := sensor.Stages[i]
+			if stage.StageGain == nil {
+				continue
+			}
+			// update the first one found, and ignore the rest
+			stage.StageGain.Value *= (r.ScaleFactor * r.GainFactor)
+			break
+		}
+	case sensor.InstrumentPolynomial != nil:
+		// First adjust for any calibrations, these are simply replacing the first two coefficients
+		if r.ScaleFactor != 1.0 || r.ScaleBias != 0.0 {
+			for i := range sensor.Stages {
+				stage := sensor.Stages[i]
+
+				// there can only be one
+				if stage.Polynomial == nil {
+					continue
+				}
+
+				// ignore changes in units, or changes in bounds
+				switch c := len(stage.Polynomial.Coefficients); {
+				case c > 1:
+					stage.Polynomial.Coefficients[1] = PolynomialCoefficient{
+						Number: stage.Polynomial.Coefficients[1].Number,
+						Value:  r.ScaleFactor,
+					}
+					stage.Polynomial.Coefficients[0] = PolynomialCoefficient{
+						Number: stage.Polynomial.Coefficients[0].Number,
+						Value:  r.ScaleBias,
+					}
+				case c > 0:
+					stage.Polynomial.Coefficients[0] = PolynomialCoefficient{
+						Number: stage.Polynomial.Coefficients[0].Number,
+						Value:  r.ScaleBias,
+					}
+				}
+
+				// update the overall instrument polynomial
+				sensor.InstrumentPolynomial.Coefficients = append([]PolynomialCoefficient{}, stage.Polynomial.Coefficients...)
```

Comment:
This may only get the last stage, is it right?

Reply:
Yes,  there can only be one polynomial stage by definition, so the last one (or none) will be fine.